src/app/views/forgotten-password/forgottenPasswordController.test.js

```diff
@@ -0,0 +1,105 @@
+        expect(component.find(".form__loader").length).toBe(0);
+    });
+});
+
+describe("When the user submits their email for a password request whilst waiting a response", () => {
+    const props = {
+        dispatch: function() {}
+    };
+    const component = mount(<ForgottenPasswordController {...props} />);
+    component.setState({
+        validationErrors: {},
+        email: {
+            value: "",
+            errorMsg: ""
+        },
+        hasSubmitted: false,
+        isSubmitting: true
+    });
+    it("Continue to show the forgottenPasswordRequest component", () => {
+        expect(component.find("h1").text()).toBe("Forgotten password");
+    });
+    it("Show the spinning loading icon", () => {
+        expect(component.find(".form__loader").length).toBe(1);
+    });
+});
+
+describe("When the user submits their email for a password request but the server returns a validation error", () => {
+    const props = {
+        dispatch: function() {}
+    };
+    const component = mount(<ForgottenPasswordController {...props} />);
+    component.setState({
+        validationErrors: {
+            heading: "Fix the following: ",
+            body: [
+                <p key="error">
+                    <a href="javascript:document.getElementById('email').focus()" className="colour--night-shadz">
+                        Enter a valid email address
+                    </a>
+                </p>
+            ]
+        },
+        email: {
+            value: "foo@bar.com",
+            errorMsg: "Enter a valid email address"
+        },
+        hasSubmitted: false,
+        isSubmitting: false
+    });
+    it("Continue to show the forgottenPasswordRequest component", () => {
+        expect(component.find("h1").text()).toBe("Forgotten password");
+    });
+    it("Stop showing the loading icon", () => {
+        expect(component.find(".form__loader").length).toBe(0);
+    });
+    it("Show the error", () => {
+        expect(component.find(".panel__error").length).toBe(1);
+        expect(component.find(".error-msg").length).toBe(1);
+    });
+});
+
+describe("When the user submits their email for a password request and the process is a success", () => {
+    const props = {
+        dispatch: function() {}
+    };
+    const component = mount(<ForgottenPasswordController {...props} />);
+    component.setState({
+        validationErrors: {},
+        email: {
+            value: "foo@bar.com",
+            errorMsg: ""
+        },
+        hasSubmitted: true,
+        isSubmitting: false
+    });
+    it("Show the forgottenPasswordEmailSent component", () => {
+        expect(component.find("h1").text()).toBe("We sent you an email");
+    });
+    it("Show the users email", () => {
+        expect(
+            component
+                .find("strong")
+                .at(0)
+                .text()
+        ).toBe("foo@bar.com");
+    });
+});
```

Comment:
I think tests also include fail scenarios for email and password

Reply:
Thanks for your review.

If the email is invalid then we don't tell the user - as this would be a security risk so the scenario is treated as a success - user is presented with a second screen saying if their email was valid then they will have an email, if they didn't get one but expected one then they need to contact an admin.

The backend can return a validation error which is handled in the test `When the user submits their email for a password request but the server returns a validation error`

The HTTP wrapper we use wraps around another wrapper (Request) that handles and logs errors, it also attempts numerous retries)

These tests are testing the react 'state' marry up to the visual elements being displayed. We have a task coming in the next sprint to add feature tests to Florence which will check each journey and scenario in more length. These feature tests will be added once the whole journey has been complete, at the moment this is only part of the journey and is disabled behind a feature flag so will not be reachable in production when released.